plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,412 @@
+#define PXI_HID_DEV_OTA_FEATURE_REPORT_ID 0x07
+
+G_DEFINE_TYPE (FuPxiDevice, fu_pxi_device, FU_TYPE_UDEV_DEVICE)
+
+
+static FuFirmware *
+fu_pxi_device_prepare_firmware (FuDevice *device,
+                                GBytes *fw,
+                                FwupdInstallFlags flags,
+                                GError **error)
+{
+    g_autoptr(FuFirmware) firmware = fu_firmware_new ();
+
+
+    guint32 idx = 0;
+
+    gsize fw_sz = g_bytes_get_size(fw);
+    guint8 pos = 0;
+    gchar version[5] = {'.', '.', '.', '.', '.'};
+    const guint8 *fw_ptr = g_bytes_get_data(fw, &fw_sz);
+
+    if (!fu_firmware_parse (firmware, fw, flags, error))
+        return NULL;
+
+    fu_common_dump_raw (G_LOG_DOMAIN, "fw last 32 bytes", &fw_ptr[fw_sz - 32], 32);
+
+    /* find the version tag */
+    for(idx = 0; idx < 32; idx++)
+    {
+        if(fw_ptr[(fw_sz - 32) + idx ] == 'v')
+        {
+            pos = idx;
+            break;
+        }
+    }
+
+    /* ensure the version */
+    if((pos != 31) && (fw_ptr[(fw_sz - 32) + pos + 1] == '_'))
+    {
+        g_debug("version in bin");
+        version[0] = fw_ptr[(fw_sz - 32) + pos + 2];
+        version[2] = fw_ptr[(fw_sz - 32) + pos + 4];
+        version[4] = fw_ptr[(fw_sz - 32) + pos + 6];
+        fu_firmware_set_version(firmware, version);
+    }
+    else
+    {
+        /* set the default version if can not find it in fw bin */
+        fu_firmware_set_version(firmware, "1.0.0");
+    }
+
+    g_debug("firmware ver %s", fu_firmware_get_version(firmware));
+
+    return g_steal_pointer (&firmware);
+}
+
+static gboolean
+fu_pxi_device_write_firmware (FuDevice *device,
+                              FuFirmware *firmware,
+                              FwupdInstallFlags flags,
+                              GError **error)
+{
+
+
+    g_autoptr(GBytes) fw = NULL;
+    g_autoptr(GArray) fw_check_sum_table = g_array_new (FALSE, TRUE, sizeof (gushort));
+    const guint8 *fw_ptr;
+    const gchar *fw_ver_ptr;
+    gsize fw_sz = 0;
+    gushort checksum = 0;
+    gushort fw_bin_checksum = 0;
+    guint32 idx = 0;
+
+    guint8 buf[PXI_OTA_BUF_SZ];
+
+    guint16 write_sz = 0;
+    guint32 offset = 0;
+    guint32 prn = 0;
+    guint16 object_size = 0;
+
+    struct cmd_fw_ota_init_new fw_ota_init_new;
+    struct cmd_fw_object_create fw_object_create;
+    struct cmd_fw_upgrade fw_upgrade;
+    struct cmd_fw_ota_disconnect fw_ota_disconnect;
+    struct hci_evt evt;
+    struct ret_fw_notify fw_notify_result;
+
+
+
+    /* get the default image */
+    fw = fu_firmware_get_image_default_bytes (firmware, error);
+    if(fw == NULL)
+        return FALSE;
+
+
+    fw_sz = g_bytes_get_size(fw);
+    fw_ptr = g_bytes_get_data(fw, &fw_sz);
+
+
+    /* calculate fw checksum */
+    fu_pxi_device_calculate_checksum(&fw_bin_checksum, fw_sz, fw_ptr);
+
+    g_debug("fw bin checksum %x, fw_sz %lu", fw_bin_checksum, fw_sz);
+    g_debug("firmware ver %s", fu_firmware_get_version(firmware));
```

Comment:
if you do the FuPxiFirmware object you can put this kind of thing in a ->to_string() vfunc.